### ISSUE TITLE
Repaired "Derelict" ships do not recharge shield/hull/energy when landed

### ISSUE THREAD

**Issue by tehhowch** [endless-sky/endless-sky#2910]
Currently, if you repair a `derelict` personality ship and then land with it at a planet with a spaceport, it will not restore any shields, hull, or energy, but will get refueled.

This seems counterintuitive, as you have already repaired it (otherwise it would not have been able to land), and every other ship type in the game is repaired upon landing at a spaceport.

The [cause of this behavior is](https://github.com/endless-sky/endless-sky/blame/master/source/Ship.cpp#L1798):
```
	if(!personality.IsDerelict())
	{
		if(atSpaceport || attributes.Get("shield generation"))
			shields = attributes.Get("shields");
		if(atSpaceport || attributes.Get("hull repair rate"))
			hull = attributes.Get("hull");
		if(atSpaceport || attributes.Get("energy generation"))
			energy = attributes.Get("energy capacity");
	}
```
The fix is to change the first line to
```
if(!personality.IsDerelict() || atSpaceport)
```
This means that when Ship::Recharge is called for any ship that has landed at a friendly spaceport, it will result in a full health ship.

A check of the existing issues about derelict ships didn't turn up any indications that the current behavior is explicitly intended, but perhaps I've missed one.
The relevant derelict / Ship::Recharge commits are
2b5bc71 - introduce derelict personality
e9ee546 - landing on uninhabited planets restores attributes for which ships have regeneration

**Comment by tehhowch** [comment]
Moving this here so it doesn't checklist the issue

<details><summary>Code Review</summary>

The Ship::Recharge method is declared with the default value of `true` for the parameter `atSpaceport`, and is called from these points in the code
 - [x] [Engine::Place, line 185](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Engine.cpp#L184)
Here the function is only called for ships which are not disabled (an unrepaired derelict ship is disabled and thus does not receive this call)
 - [x] [Engine::CalculateStep, line 1540](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Engine.cpp#L1540)
Here the function is called for a special `person` that has been selected to appear in the game. This call to Recharge precedes the call to SetPersonality, which is the method that would make a derelict ship actually become a derelict, disabled ship (i.e. not applicable).
 - [x] [OutfitterPanel::Buy, line 337](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/OutfitterPanel.cpp#L337)
Called when you buy outfits for a ship (i.e. not applicable).
 - [x] [OutfitterPanel::Sell, line 531](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/OutfitterPanel.cpp#L531)
Called when you sell outfits for a ship (i.e. not applicable).
 - [x] [PlayerInfo::Land, line 943](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/PlayerInfo.cpp#L943)
This call is only performed for ships that the player owns (i.e. not applicable).
 - [x] [PlayerInfo::TakeOff, line 1102](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/PlayerInfo.cpp#L1100)
This call is only performed for ships that the player owns (i.e. not applicable).
 - [x] [Ship::Restore, line 1771](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Ship.cpp#L1771)
Ship::Restore is called only from GameData::Revert(), e.g. when loading a savegame, and only for the `person` ships.
 - [x] [Ship::FinishLoading, line 367](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Ship.cpp#L367)
This is primarily called from GameData::BeginLoading, in which the game constructs what each ship is, but is also called by NPC::Load (just before SetPersonality is called, which will override it), and NPC::Instantiate (again, before SetPersonality is called).

This indicates that there are no unintended side effects of making the specified change.

The other call of concern is SetPersonality, as it is the method that effectively cripples a derelict ship.
Making the changes to Recharge() will not prevent a later call to SetPersonality from overriding them:
SetPersonality is called in these locations
 - [x] [Engine::CalculateStep, line 1543](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Engine.cpp#L1543)
As above, this is for a special person being instantiated into the game, and should not be altered.
 - [x] [Fleet::Instantiate line 414](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Fleet.cpp#L414)
This is called when a group of non-special ships is spawned - i.e. not mission NPCs.
 - [x] [NPC::Load, line 141](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/NPC.cpp#L141)
This is called when the game is loaded from a savegame, and needs updating - if you repair a derelict ship, then land and quit the game, you will have to repair the derelict again upon takeoff. However, this is not related to the changes made to Ship::Recharge()
 - [x] [NPC::Instantiate, line 397](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/NPC.cpp#L397)
This is called when creating the mission NPC ships, and does not need any modification (this creates the initial derelict condition, before you could have had a chance to repair it).
 - [x] [Planet::DeployDefense, line 523](https://github.com/endless-sky/endless-sky/blob/2e813777c01e36af5f6e5d35caa2056d7193432d/source/Planet.cpp#L523)
(not relevant)
</details>

**Comment by endless-sky** [comment]
Fixed in ee9d339.
